firmware/apps/nic/pv.uc

```diff
@@ -505,6 +515,113 @@ end#:
 .end
 #endm
 
+#macro pv_get_xdp_meta_len(out_len, in_vec, RET_LABEL)
+.begin
+
+    .reg xml
+
+    br_bclr[BF_AL(in_vec, PV_XDP_META_bf), RET_LABEL], defer[1]
+        immed[out_len, 0]
+
+    bitfield_extract__sz1(xml, BF_AML(in_vec, PV_XDP_META_LEN_bf))
+    alu[out_len, xml, +, 1]
+    alu[out_len, --, B, out_len, <<2]
+
+.end
+#endm
+
+#macro pv_save_xdp_meta(xdp_meta_len, in_vec)
+.begin
+    .reg offset
+    .reg lm_ptr
+    .reg in_pkt_addr_hi
+    .reg in_pkt_addr_lo
+    .reg xdp_meta_base
+    .reg read $tr
+    .sig sig_rd
+
+    pv_get_xdp_meta_len(xdp_meta_len, in_vec, end#)
+    pv_get_base_addr(in_pkt_addr_hi, in_pkt_addr_lo, in_vec)
+    alu[xdp_meta_base, in_pkt_addr_lo, -, xdp_meta_len]
+
+    local_csr_rd[ACTIVE_LM_ADDR_1]
+    immed[lm_ptr, 0]
+    alu[lm_ptr, lm_ptr, +, 32]
+    local_csr_wr[ACTIVE_LM_ADDR_2, lm_ptr]
+
+    alu[offset, 32, -, xdp_meta_len] // jump offset
+    jump[offset, t8#], targets[t8#, t7#, t6#, t5#, t4#, t3#, t2#, t1#]
+
+#define LOOP 8
+#while (LOOP > 1)
+
+t/**/LOOP#:
+    mem[read32, $tr, in_pkt_addr_hi, <<8, xdp_meta_base, 1], ctx_swap[sig_rd]
```

Comment:
I am guessing we do not have enough xfer registers to indirect_ref xdp_meta_base read this in a block?
This must be the case since this macro is called from a block of code that calls pv_invalidate_cache

Reply:
Yes, this is the case. What is the right way to fix this?

Reply:
I think you can to re-use $__pv_pkt_data[32]. This is used to cache packet header data. When you use pv_save_xdp_meta in ebf_reentry, you can do pv_invalidate_cache(in_vec) afterward. On the next pv_seek the packet header will be re-read.